### ISSUE TITLE
Should the compiler error if no modules in a source file match the file's name?

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#21702]
There is a test `modules/implicit/parseTwice/main.chpl` that tries
to import a module that does not match its filename. The test is
small, and the code looks like:

```chapel
// SomeLibrary.chpl
module M {
  var x = 42;
}

// main.chpl
proc main() {
  use SomeLibrary only;
  writeln(SomeLibrary.M.x);
}
```

The error looks like (`-->` are my attributions):

```
--> This first line is added by the new frontend:
/SomeLibrary.chpl:1: error: In use/imported file, module name M
does not match file name SomeLibrary.chpl
--> Old error appears from both production and dyno...
main.chpl:1: In function 'main':                                                                                                                                                     │
main.chpl:2: error: cannot find module or enum named 'SomeLibrary'
```

Both the production compiler and dyno emit an error about how no
module named `SomeLibrary` was able to be found.

However dyno also emits an error when the AST of `SomeLibrary.chpl`
is inspected, saying that the file name does not match the name of
any modules found inside it.

Is this error (or some version of it) worth keeping?

I think this new error is useful because it helps give some context
for the `use`/`import` error that follows.

It seems like given a file named `SomeLibrary.chpl`, the expectation
that there is a module named `SomeLibrary` ought to be honored
more closely.

**Comment by dlongnecke-cray** [comment]
One thing I'm worried about with this error is that we have rules for when an implicit module is created (I'm not entirely certain of the rules, but know you can get warnings when implicit modules are created in cases where you may not have expected it).

**Comment by lydia-duncan** [comment]
I don't think we should require the presence of a module that matches the file name.  It's been extremely useful to be able to write multiple test files in the same directory with the same module name in their body, to track down variations of an error.

I do think the additional information when a user tries to use the name that would be generated is helpful.  But I wouldn't go further than that

**Comment by dlongnecke-cray** [comment]
One thing that's interesting is that this error isn't really that common - among all the tests we have it was only parse-twice that seemed to run into this error as far as I know, and the actual bug was that it was emitting the error too many times.

@lydia-duncan: I've edited the title to be more accurate. I expect that this error would only fire on use/import, not in general if you have a `Foo.chpl` without a `Foo` module inside it.

**Comment by bradcray** [comment]
I asked myself a similar question to this when reviewing https://github.com/chapel-lang/chapel/pull/21596 today.  Here are my thoughts:

* Like Lydia, I don't think it's reasonable to always complain when a filename contains a module by a different name.  In addition to the case she cites, I think it's reasonable to have a file named `M.chpl` defining a module M and then to create code clones of it over time, like `M-2.0.chpl`, `M-bug.chpl`, etc. and be able to compile them without getting an error or warning
* In the (one?) case I saw in #21596, the warning happened when the file in question was not named on the `chpl` command-line, but was dragged into the compilation by virtue of a `use` statement.  E.g., a `use N;` causes `N.chpl` to be parsecd, but finds that the file doesn't define a module named `N` but perhaps some other module(s) like `O` instead.  Again, this doesn't seem like an obviously erroneous condition to me, but it does seem potentially confusing/surprising enough to be worth issuing an error for.  And since it looked like only one test was modified in this way in #21596, it doesn't seem to trigger in common cases / use patterns (as I'd expect).

So basically, I decided it was a reasonable warning to issue for an implicitly parsed `.chpl` file, but wouldn't want to see it generated (by default) for an explicitly-parsed one (i.e., one named on the command-line).  (I could imagine an opt-in warning or lint-style check for it for people who wanted to verify that a given set of files were pristine).

Oops, after writing all the above, I see that the test in question on #21596 actually does generate an error rather than a warning.  To me that feels unnecessarily heavy-handed if the module ends up happening to be resolved in some other way.  I don't mean to suggest that this will support beautiful/natural/preferred code patterns, simply that it doesn't seem like it should be an error to me to have a file not define a module whose name matches its filename.  Tagging @DanilaFe due to the interaction with #21596.

**Comment by mppf** [comment]
Right, I've argued in the past that the module name and file name should match in general, but that's not what we're talking about here. In this issue, we're talking about *the specific case of a `use M` that finds `M.chpl` that contains some module other than `M`*.

I don't understand how anything can make sense in this situation. We've read in a module that is not the module that was requested. Should we just ignore result?

> * Again, this doesn't seem like an obviously erroneous condition to me, but it does seem potentially confusing/surprising enough to be worth issuing an error for.

I don't understand how compilation can proceed.

It doesn't work with the production compiler:

``` chapel
// SomeLibrary.chpl
module M {
  var x = 42;
}
```

``` chapel
// main.chpl
use SomeLibrary;
writeln(x);
```

What we get in production is this:

``` chapel
$ chpl main.chpl 
main.chpl:1: error: Cannot find module or enum 'SomeLibrary'
```

It seems like a strict improvement over this error to also say that "The reason we can't find 'SomeLibrary' is that we found SomeLibrary.chpl but it did not contain that module".

However, I think that the existing dyno error wording

> In use/imported file, module name M does not match file name SomeLibrary.chpl 

is throwing everybody off, and we need different wording to keep this error.

Here is an attempt:

> A use or import statement caused the compiler to look for a module SomeLibrary in SomeLibrary.chpl according to the module search path. It found SomeLibrary.chpl but that file does not contain a module named SomeLibrary.

**Comment by bradcray** [comment]
Sorry to be unclear:  What I was imagining was a warning followed by an error:

> SomeLibrary.chpl:1: warning: In used/imported file, module name 'M' does not match file name 'SomeLibrary.chpl'
> main.chpl:1: error: Cannot find module or enum 'SomeLibrary'

where my hope is to indicate "It was surprising to me that I didn't find 'SomeLibrary' in the implicitly compiled file named 'SomeLibrary.chpl' (but that that in itself was not an error).  But then I couldn't find the module that you wanted to include, and that is an error."

My thinking is:
* the combination of both messages seems strictly better than either in isolation
* it seems plausible that something else other than `SomeLibrary.chpl` _might_ provide the 'SomeLibrary' module before its symbol is resolved (even though this is unlikely in any well-structured program, and potentially isn't possible in either of the current compiler architectures, maybe it would be in some future approach).

**Comment by mppf** [comment]
That sounds reasonable in the near term, because the `Cannot find module or enum 'SomeLibrary'` error is emitted by the production compiler, but longer-term, I think we might be better served trying to get a single, informative error message here.

> it seems plausible that something else other than `SomeLibrary.chpl` _might_ provide the 'SomeLibrary' module before its symbol is resolved (even though this is unlikely in any well-structured program, and potentially isn't possible in either of the current compiler architectures, maybe it would be in some future approach).

Are you able to come up with an example that does that that we could discuss? (Even if it does not compile today)?

**Comment by bradcray** [comment]
> Are you able to come up with an example that does that that we could discuss?

Not that's code I'd be willing to write.  Basically the idea would be "compiler collects all module/enum names in 'use' statements; for all such names that the compiler can't already resolve, it parses all filenames whose names match those symbols; one of those files doesn't contain the  module we expected/hoped it would, but another does (e.g., maybe SomeOtherLibrary.chpl defines both SomeLibrary and SomeOtherLibrary); the compiler still warns that this code is wonky, but resolves both symbols and continues compiling.

All that said, my main argument for using a warning rather than an error is not to support cases or compiler architectures like this, but to get the two distinct messages making it clear that the first is surprising/weird/alarming/worth improving, but not technically illegal while the other is the true error and reason we can't/won't compile the program.

**Comment by mppf** [comment]
> > Are you able to come up with an example that does that that we could discuss?

> Not that's code I'd be willing to write. 

Maybe I am less scrupulous?

Here is such an example.

``` chapel
// main.chpl
module Main {
  proc main() {
    use SomeLibrary;
    use OtherLibrary;
    writeln(x);
  }
}
```

``` chapel
// SomeLibary.chpl
module M {
  var x = 42;
}
```

``` chapel
// OtherLibrary.chpl
module OtherLibrary { }
module SomeLibrary { var x = "boo"; }
```

It compiles and runs in production.

```
$ chpl main.chpl 
$ ./main
boo
```

When using the dyno scope resolver, it does not compile.

```
$ chpl --dyno main.chpl 
./SomeLibrary.chpl:1: error: In use/imported file, module name M does not match file name SomeLibrary.chpl
main.chpl:2: In function 'main':
main.chpl:3: error: cannot find module or enum named 'SomeLibrary'
```

In any case, I am understanding that we are in agreement that this example should not compile, going forward.

**Comment by bradcray** [comment]
> is it your viewpoint that emitting a separate warning and error has educational value to users, beyond one error message that is more descriptive than it is today?

A major reason I'd prefer the "`M.chpl` didn't define a module named `M`" message to be a warning rather than an error is that I don't think the language or language spec should require that a filename and module name must match, and I tend to think of errors as being strongly correlated to how the language is defined / what the spec says.

> In any case, I am understanding that we are in agreement that this example should not compile, going forward.

No, I'd actually be fine if your example continued to compile going forward, though I'm not at all insistent that it continue to compile.  Again, it's not code I'd write or be proud of in any way, but if the production compiler were going forward, I wouldn't change the production compiler to stop supporting that case.

Making sure I understand this:

> Because that is the situation I am expecting we will end up with when we get a bit further with migrating to the dyno scope resolver.

Is the point that dyno will be more depth-first by nature, so in processing the `use` and failing to find the symbol, it will search for the file, load it, try to resolve the symbol again, still not find it, so be unable to proceed?  If that's the case, I have no problem if your program didn't continue to work.  If it's something else, I'd like to reserve judgement until understanding it better.

> ```
> warn("In used/imported file, module name '%s' does not match file name '%s'", ...);
> error("Cannot find module or enum '%s'", ...);
> ```

I suppose that would be tolerable to me, though it doesn't seem nearly as attractive as having the scope resolver fail to find the symbol in the normal way (say, if there were no file named M.chpl to use to resolve the  missing 'M').

**Comment by mppf** [comment]
> > Because that is the situation I am expecting we will end up with when we get a bit further with migrating to the dyno scope resolver.
> 
> Is the point that dyno will be more depth-first by nature, so in processing the `use` and failing to find the symbol, it will search for the file, load it, try to resolve the symbol again, still not find it, so be unable to proceed?

Yes, that's my understanding. It doesn't really have a way to work with global state the way the production compiler does. Anyway, I think it's highly likely that we can make variations of https://github.com/chapel-lang/chapel/issues/21702#issuecomment-1453625500 by switching around where the `module SomeLibrary` appears so that it doesn't compile even in production (due to order of things happening in the compiler). I'd rather simply not allow code like this than to try to explain (e.g. in the spec) when or why it is allowed.

> > In any case, I am understanding that we are in agreement that this example should not compile, going forward.
> 
> No, I'd actually be fine if your example continued to compile going forward, though I'm not at all insistent that it continue to compile. Again, it's not code I'd write or be proud of in any way, but if the production compiler were going forward, I wouldn't change the production compiler to stop supporting that case.

Thanks for clarifying. I will re-summarize. We are in agreement that it is OK if this example does not compile.
